File: ibeacon/Managers/BeaconManager.m
//
//  BeaconManager.m
//  ibeacon
//
//  Created by Migele Beloded on 9/26/14.
//  Copyright (c) 2014 obrichak. All rights reserved.
//

#import "BeaconManager.h"

NSString * const DidEnterRegionNotification = @"DidEnterRegionNotification";
NSString * const DidExitRegionNotification = @"DidExitRegionNotification";
NSString * const RangingNotification = @"RangingNotification";

@interface BeaconManager()

@property (nonatomic, strong) CLLocationManager *locationManager;

@end

@implementation BeaconManager

+ (instancetype)sharedManager {
	static BeaconManager *_sharedManager;
	
	if (!_sharedManager) {
		_sharedManager = [[[self class] alloc] init];
		NSBlockOperation *op = [NSBlockOperation blockOperationWithBlock:^{
			CLLocationManager *locationManager = [[CLLocationManager alloc] init];
			locationManager.delegate = _sharedManager;
			//These two settings sacrifice battery life for a smoother demo experience.
			[locationManager disallowDeferredLocationUpdates];
			locationManager.pausesLocationUpdatesAutomatically = NO;
			
			_sharedManager->_locationManager = locationManager;
		}];
		[[NSOperationQueue mainQueue] addOperation:op];
		[op waitUntilFinished];
		
	}
	return _sharedManager;
}

- (void)beginRegionMonitoring
{
//	NSArray *regionKeys;
//	@synchronized(self) {
//		regionKeys = [self.URLStore allKeys];
//	}
//	for (NSString *key in regionKeys) {
//		NSArray *regionTriplet = [key componentsSeparatedByString:@":"];
//		NSUUID *proximityUUID = [[NSUUID alloc] initWithUUIDString:regionTriplet[0]];
//		CLBeaconMajorValue major = [(NSString *)regionTriplet[1] integerValue];
//		CLBeaconMinorValue minor = [(NSString *)regionTriplet[2] integerValue];
//		[[NSOperationQueue mainQueue] addOperationWithBlock:^{
//			CLBeaconRegion *region = [[CLBeaconRegion alloc] initWithProximityUUID:proximityUUID major:major minor:minor identifier:key];
//			[self.locationManager startMonitoringForRegion:region];
//		}];
//	}
	
}

#pragma mark - CLLocationManager Delegate

- (void)locationManager:(CLLocationManager *)manager didEnterRegion:(CLRegion *)region
{
	NSLog(@"Entered Region");
	UIApplicationState state = [[UIApplication sharedApplication] applicationState];
	switch (state) {
		case UIApplicationStateInactive:
		case UIApplicationStateActive:
			[[NSNotificationCenter defaultCenter] postNotificationName:DidEnterRegionNotification
																object:self
															  userInfo:@{@"region": region}];
            
			break;
		case UIApplicationStateBackground: {
			UILocalNotification *notification = [[UILocalNotification alloc] init];
//			notification.alertBody = self.entryText;
			[[UIApplication sharedApplication] presentLocalNotificationNow:notification];
		}
		default:
			break;
	}
    
}

- (void)locationManager:(CLLocationManager *)manager didExitRegion:(CLRegion *)region
{
	NSLog(@"Exited Region");
	UIApplicationState state = [[UIApplication sharedApplication] applicationState];
	switch (state) {
		case UIApplicationStateInactive:
		case UIApplicationStateActive:
			[[NSNotificationCenter defaultCenter] postNotificationName:DidExitRegionNotification
																object:self
															  userInfo:@{@"region": region}];
            
			break;
		case UIApplicationStateBackground:{
			UILocalNotification *notification = [[UILocalNotification alloc] init];
//			notification.alertBody = self.exitText;
			[[UIApplication sharedApplication] presentLocalNotificationNow:notification];
		}
		default:
			break;
	}
	[manager startMonitoringForRegion:region];
}

- (void)locationManager:(CLLocationManager *)manager didRangeBeacons:(NSArray *)beacons inRegion:(CLBeaconRegion *)region
{
	NSLog(@"Ranged %lu Beacons", (unsigned long)[beacons count]);
	if ([beacons count] > 0) {
		//Doesn't deal gracefully with multiple beacons in range
		CLBeacon *closestBeacon = beacons[0];
		NSNumber *index;
		switch (closestBeacon.proximity) {
			case CLProximityFar:
				index = @0;
				break;
			case CLProximityNear:
				NSLog(@"Near");
				index = @1;
				break;
			case CLProximityImmediate:
				NSLog(@"Immediate");
				index = @2;
				break;
			case CLProximityUnknown:
				NSLog(@"Unknown");
				index = @3;
				break;
		}
		[[NSNotificationCenter defaultCenter] postNotificationName:RangingNotification object:self userInfo:@{@"index": index, @"beacon": closestBeacon}];
	}
}

- (void)locationManager:(CLLocationManager *)manager didDetermineState:(CLRegionState)state forRegion:(CLRegion *)region{
	CLBeaconRegion *beaconRegion = (CLBeaconRegion *)region;
	switch (state) {
		case CLRegionStateInside:
			NSLog(@"Determined Inside State for Beacon %@", beaconRegion.identifier);
			if ([CLLocationManager isRangingAvailable]) {
				[self.locationManager startRangingBeaconsInRegion:beaconRegion];
			}
			break;
		case CLRegionStateOutside:
			NSLog(@"Determined Outside State for Beacon %@", beaconRegion.identifier);
			if ([CLLocationManager isRangingAvailable]) {
				[self.locationManager stopRangingBeaconsInRegion:beaconRegion];
			}
			break;
		case CLRegionStateUnknown:
			NSLog(@"Determined Unknown State for Beacon %@", beaconRegion.identifier);
			break;
	}
}

@end

File: PassKitApp/PassKitApp/Classes/Managers/BeaconManager/BeaconManager.swift
//
//  BeaconManager.swift
//  PassKitApp
//
//  Created by Migele Beloded on 10/6/14.
//  Copyright (c) 2014 Alexandr Chernyy. All rights reserved.
//

import Foundation
protocol BeaconManagerDelegate
{

}

class BeaconManager: NSObject, BluetoothManagerDelegate, CLLocationManagerDelegate {
    
    private let proximityUUID = NSUUID(UUIDString:"19D5F76A-FD04-5AA3-B16E-E93277163AF6")
    private let locationManager     : CLLocationManager!
    private var beaconRegion        : CLBeaconRegion!
    private var btStatus            : String!
    private var delegate            : BeaconManagerDelegate?
    private var isBeaconAchieved    : Bool
    
    class var sharedInstance : BeaconManager {
        
    struct Static {
        static var onceToken : dispatch_once_t = 0
        static var instance : BeaconManager? = nil
        }
        dispatch_once(&Static.onceToken) {
            Static.instance = BeaconManager()
        }
        return Static.instance!
    }
    
    // PRIVATE
    override init() {
        
        self.isBeaconAchieved = false;
        
        super.init()
        
        locationManager = CLLocationManager()
        
        setup()
        
    }
    
    func setup()
    {
        beaconRegion = CLBeaconRegion(proximityUUID:proximityUUID,identifier:"GemTot USB")
        
        BluetoothManager.sharedInstance
        BluetoothManager.sharedInstance.delegate = self
        
        locationManager.delegate = self
        locationManager.desiredAccuracy = kCLLocationAccuracyBest
        locationManager.startUpdatingLocation()
    }
    
//LocationManager Delegate
    func locationManager(manager: CLLocationManager!, didStartMonitoringForRegion region: CLRegion!) {
        manager.requestStateForRegion(region)
        println("Scanning...")
    }
    
    func locationManager(manager:CLLocationManager, didUpdateLocations locations:[AnyObject]) {
        println("locations = \(locations)")
    }
    
    func locationManager(manager: CLLocationManager!, didEnterRegion region:CLRegion!) {
        if region.isKindOfClass(CLBeaconRegion)
        {
            println("HI")
            AlertManager.sharedInstance.showAlert("iBeacon", msg: "HI", delegate: nil, btn: "Ok")
        }
    }
    
    func locationManager(manager: CLLocationManager!, didExitRegion region:CLRegion!) {
        if region.isKindOfClass(CLBeaconRegion)
        {
            if self.isBeaconAchieved {
                self.isBeaconAchieved = false
                println("GOOD_BYE")
                AlertManager.sharedInstance.showAlert("iBeacon", msg: "GOOD_BYE", delegate: nil, btn: "Ok")
            }
        }
    }
    
    func locationManager(manager: CLLocationManager!, monitoringDidFailForRegion region: CLRegion!, withError error: NSError!) {
//        NSLog(@"Failed monitoring region: %@", error);
//        [alertManager showAlert:@"Failed monitoring region" errorMsg:error.description];
        println("Failed monitoring region: \(error)")
    }

    func locationManager(manager: CLLocationManager!, didFailWithError error:NSError!) {
        println("Location manager failed: \(error)");
//        [alertManager showAlert:@"Location manager failed: " errorMsg:error.description];
    }

    func locationManager(manager: CLLocationManager!, didRangeBeacons beacons:NSArray!, inRegion region:CLBeaconRegion!)
    {
//        var beacon:CLBeacon
//        
//        for beacon in beacons {
////            for (BeaconModel item in self.items) {
////                if ([item isEqualToCLBeacon:beacon]) {
////                    item.lastSeenBeacon = beacon;
////                    [self showBeaconInfo:beacon];
////                }
////            }
//        
//        }
        println(beacons)
        
        if(beacons.count == 0) { return }

        var beacon = beacons[0] as CLBeacon
        
        /*
        beacon
        proximityUUID   :   region name
        major           :   id１
        minor           :   id２
        proximity       :   provim
        accuracy        :   acc
        rssi            :   rssi
        */
        if (beacon.proximity == CLProximity.Unknown) {
            println("Unknown Proximity")
//            reset()
            return
        } else if (beacon.proximity == CLProximity.Immediate) {
            println("Immediate")
        } else if (beacon.proximity == CLProximity.Near) {
            println("Near")
        } else if (beacon.proximity == CLProximity.Far) {
            println("Far")
        }
    }
    
    func locationManager(manager: CLLocationManager!, didDetermineState state: CLRegionState, inRegion: CLBeaconRegion!)
    {
        if (state == .Inside) {
            manager.startRangingBeaconsInRegion(inRegion)
            AlertManager.sharedInstance.showAlert("beacon info", msg: "You are inside of beacon range", delegate: nil, btn: "Ok")
        }

//        // Scan our registered notifications to see if this state change
//        // has a user message associated with it
//        NSString * stateStr = @"";
//        
//        switch (state) {
//        case CLRegionStateUnknown:
//        {
//        [manager stopRangingBeaconsInRegion:(CLBeaconRegion*)region];
//        stateStr = @"unknown";
//        break;
//        }
//        case CLRegionStateInside:
//        {
//        [manager startRangingBeaconsInRegion:(CLBeaconRegion*)region];
//        stateStr = @"You are inside of beacon range";
//        break;
//        }
//        case CLRegionStateOutside:
//        {
//        [manager stopRangingBeaconsInRegion:(CLBeaconRegion*)region];
//        stateStr = @"You are outside of beacon range";
//        if(isBeaconAchieved) {
//        isBeaconAchieved = NO;
//        [self.startViewController hideAll];
//        }
//        break;
//        
//        default:
//        break;
//        }
//        }
//        [self.startViewController.rangeStatus setText:stateStr];
    
    }
    
    func startMonitoringItem(/*item: BeaconModel*/)
    {
    
        switch CLLocationManager.authorizationStatus() {
        case .Authorized, .AuthorizedWhenInUse:

            self.locationManager.startRangingBeaconsInRegion(self.beaconRegion)
        case .NotDetermined:
            AlertManager.sharedInstance.showAlert("iBeacon", msg: "RStarting Monitor", delegate: nil, btn: "Ok")
//            if(UIDevice.currentDevice().systemVersion.substringToIndex(1).toInt() >= 8){
//
//                self.locationManager.requestAlwaysAuthorization()
//            }else{
//                self.locationManager.startRangingBeaconsInRegion(self.beaconRegion)
//            }
        case .Restricted, .Denied:

            println("Restricted")
            AlertManager.sharedInstance.showAlert("iBeacon", msg: "Restricted Monitor", delegate: nil, btn: "Ok")
        }
//        self.beaconRegion = [self beaconRegionWithItem:item];
//
//        [self.locationManager startMonitoringForRegion:self.beaconRegion];
//        [self.locationManager startRangingBeaconsInRegion:self.beaconRegion];
//        
//        [[UIApplication sharedApplication] cancelAllLocalNotifications];
    
    }
    
    func stopMonitoringItem()
    {
//        if(!self.beaconRegion)
//        return;
//        
//        [self.locationManager stopMonitoringForRegion:self.beaconRegion];
//        [self.locationManager stopRangingBeaconsInRegion:self.beaconRegion];
    }
    
//Bluetooth Delegate
    func bluetoothReady()
    {
        println("Start of beacons scanning")
//        AlertManager.sharedInstance.showAlert("Bluetooth", msg: "Start of beacons scanning", delegate: nil, btn: "Ok")
        startMonitoringItem()
    }
    
    func bluetoothError(error:String)
    {
        println("Error: \(error)")
//        AlertManager.sharedInstance.showAlert("Bluetooth", msg: "Error: \(error)", delegate: nil, btn: "Ok")
        stopMonitoringItem()
    }
    
}